File: app/com/lynxanalytics/biggraph/graph_operations/PulledOverEdges.scala
// This operation is useful when one wants to apply on an edge bundle a tranformative operation
//  (e.g. filter) that was defined on vertex sets.
//
// The way to go is to first apply the operation on edgeBundle.idSet, say it creates a new
// vertex set destinationVS and also provides an injection from destinationVS to
// edgeBundle.idSet. Then apply PullOverEdges with the original edge bundle and the above
// injection.

package com.lynxanalytics.biggraph.graph_operations

import com.lynxanalytics.biggraph.graph_api._
import com.lynxanalytics.biggraph.spark_util.Implicits._

object PulledOverEdges extends OpFromJson {
  class Input extends MagicInputSignature {
    val originalSrc = vertexSet
    val originalDst = vertexSet
    val originalIds = vertexSet
    val originalEB = edgeBundle(originalSrc, originalDst, idSet = originalIds)
    val destinationVS = vertexSet
    val injection = edgeBundle(destinationVS, originalIds, EdgeBundleProperties.injection)
  }
  class Output(implicit instance: MetaGraphOperationInstance,
               inputs: Input) extends MagicOutput(instance) {
    val pulledEB = edgeBundle(
      inputs.originalSrc.entity,
      inputs.originalDst.entity,
      idSet = inputs.destinationVS.entity)
  }
  def fromJson(j: JsValue) = PulledOverEdges()
}
import PulledOverEdges._
case class PulledOverEdges()
    extends TypedMetaGraphOp[Input, Output] {
  override val isHeavy = true
  @transient override lazy val inputs = new Input()

  def outputMeta(instance: MetaGraphOperationInstance) = new Output()(instance, inputs)

  def execute(inputDatas: DataSet,
              o: Output,
              output: OutputBuilder,
              rc: RuntimeContext): Unit = {
    implicit val id = inputDatas
    val injectionEntity = inputs.injection.meta
    val injection = inputs.injection.rdd
    val originalEB = inputs.originalEB.rdd
    val destinationVS = inputs.destinationVS.rdd
    val destinationPartitioner = destinationVS.partitioner.get
    val pulledEB =
      if (injectionEntity.properties.isIdPreserving) {
        val joinableOriginalEB = originalEB.sortUnique(destinationPartitioner)
        joinableOriginalEB.sortedJoin(destinationVS).mapValues { case (edge, _) => edge }
      } else {
        val originalToDestinationId = injection
          .map { case (id, edge) => (edge.dst, edge.src) }
          .sortUnique(originalEB.partitioner.get)
        originalEB.sortedJoin(originalToDestinationId)
          .map { case (originalId, (edge, destinationId)) => (destinationId, edge) }
          .sortUnique(destinationPartitioner)
      }
    output(o.pulledEB, pulledEB)
  }
}

File: sphynx/server/pulled_over_edges.go
// Implements the PulledOverEdges operation
// See the Spark implementation for details

package main

func doPulledOverEdges(
	destinationVS *VertexSet,
	originalEB *EdgeBundle,
	injection *EdgeBundle) *EdgeBundle {

	origIds := make(map[int]int, len(injection.Dst))
	for i := 0; i < len(injection.Src); i++ {
		origIds[injection.Src[i]] = injection.Dst[i]
	}
	pulledEB := &EdgeBundle{
		Src:         make([]int, len(injection.Src)),
		Dst:         make([]int, len(injection.Src)),
		EdgeMapping: make([]int64, len(injection.Src)),
	}
	j := 0
	for i := 0; i < len(destinationVS.MappingToUnordered); i++ {
		origId, exists := origIds[i]
		if exists {
			pulledEB.Src[j] = originalEB.Src[origId]
			pulledEB.Dst[j] = originalEB.Dst[origId]
			pulledEB.EdgeMapping[j] = originalEB.EdgeMapping[origId]
			j++
		}
	}
	return pulledEB
}

func init() {
	operationRepository["PulledOverEdges"] = Operation{
		execute: func(ea *EntityAccessor) error {
			destinationVS := ea.getVertexSet("destinationVS")
			originalEB := ea.getEdgeBundle("originalEB")
			injection := ea.getEdgeBundle("injection")
			pulledEB :=
				doPulledOverEdges(destinationVS, originalEB, injection)
			ea.output("pulledEB", pulledEB)
			return nil
		},
	}
}
